sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -310,6 +310,31 @@ class CodegenContext {
       // The UTF8String may came from UnsafeRow, otherwise clone is cheap (re-use the bytes)
       case StringType => s"$row.update($ordinal, $value.clone())"
       case udt: UserDefinedType[_] => setColumn(row, udt.sqlType, ordinal, value)
+      case s: StructType if UnsafeRow.isMutable(s) =>
+        val nestedRow = freshName("nestedRow")
+        val updateFields = s.zipWithIndex.map { case (field, index) =>
+          val ev = ExprCode(
+            code = "",
+            isNull = s"value.isNullAt($index)",
+            value = getValue("value", field.dataType, index.toString))
+          updateColumn(nestedRow, field.dataType, index, ev, field.nullable)
+        }
+        val updateFieldsCode = splitExpressions(
+          updateFields, "updateFields", Seq("InternalRow" -> nestedRow, "InternalRow" -> "value"))
+        val setColumnFunc = freshName("setColumnFunc")
+        val funcCode = s"""
+          public void $setColumnFunc(InternalRow row, InternalRow value) {
+            if (row instanceof UnsafeRow) {
+              ((UnsafeRow) row).setNotNullAt($ordinal);
+              final InternalRow $nestedRow = row.getStruct($ordinal, ${s.length});
+              $updateFieldsCode
```

Comment:
> In that case the value row and the nestedRow should be the same

yea we should skip updating for this case.

> we could just copy the bytes from value to nestedRow.

well, we can't guarantee the `value` is `UnsafeRow` and fixed-length